File: scripts/CV.py
# coding: utf-8
# CV.py

from __future__ import annotations
from itertools import islice
from typing import Any
import random

from common import read_csv, write_csv


#################### CrossValidation ####################

class CrossValidation:
	def __init__(self, samples: list[str], table: list[list[int]]):
		self.samples: list[str] = samples
		self.CV: list[list[int]] = table
	
	def num_repeats(self):
		return len(self.CV)
	
	def num_divisions(self):
		return max(self.CV[0])
	
	def extract_X_train(self, X: list[list[Any]],
								k: int, g: int) -> list[list[Any]]:
		v = self.CV[k]
		return [ [ x for x, g1 in zip(row, v) if g1 != g ] for row in X ]
	
	def extract_y_train(self, y: list[float], k: int, g: int) -> list[float]:
		v = self.CV[k]
		return [ y1 for y1, g1 in zip(y, v) if g1 != g ]
	
	def extract_X_test(self, X: list[list[Any]],
								k: int, g: int) -> list[list[Any]]:
		v = self.CV[k]
		return [ [ x for x, g1 in zip(row, v) if g1 == g ] for row in X ]
	
	def set_values(self, pred: list[float], p_sub: list[float], k: int, g: int):
		v = self.CV[k]
		for i, value in zip((i for i, g1 in enumerate(v) if g1 == g), p_sub):
			pred[i] = value
	
	def write(self, path: str):
		nrep = self.num_repeats()
		table = [[''] + [ 'CV%d' % i for i in range(1, nrep+1) ]]
		for s, *v in zip(self.samples, *self.CV):
			table.append([s] + list(v))
		write_csv(table, path)
	
	@staticmethod
	def create_randomly(samples: list[str],
						ndiv: int, nrep: int) -> CrossValidation:
		N = len(samples)
		CV: list[list[int]] = []
		for i in range(nrep):
			v = [ g+1 for g in range(ndiv)
						for _ in range(g*N//ndiv, (g+1)*N//ndiv) ]
			random.shuffle(v)
			CV.append(v)
		return CrossValidation(samples, CV)
	
	@staticmethod
	def read(path: str) -> CrossValidation:
		table = list(read_csv(path))
		samples = [ v[0] for v in table[1:] ]
		CV = [ [ int(table[j][i]) for j in range(1, len(table)) ]
										for i in range(1, len(table[0])) ]
		return CrossValidation(samples, CV)

File: src/CV.cpp
// CV.cpp

#include <string>
#include <iostream>
#include <fstream>
#include "../include/CV.h"
#include "../include/common.h"

using namespace std;

vector<vector<double>> CrossValidation::extract_X_train(
									const vector<vector<double>>& X,
									size_t k, int g) {
	const vector<int>&	v = table[k];
	vector<vector<double>>	new_X;
	for(size_t i = 0; i < X.size(); ++i) {
		vector<double>	row;
		for(size_t j = 0; j < X[i].size(); ++j) {
			if(v[j] != g)
				row.push_back(X[i][j]);
		}
		new_X.push_back(row);
	}
	return new_X;
}

vector<double> CrossValidation::extract_y_train(const vector<double>& y,
															size_t k, int g) {
	const vector<int>&	v = table[k];
	vector<double>	new_y;
	for(size_t i = 0; i < y.size(); ++i) {
		if(v[i] != g)
			new_y.push_back(y[i]);
	}
	return new_y;
}

vector<vector<double>> CrossValidation::extract_X_test(
									const vector<vector<double>>& X,
									size_t k, int g) {
	const vector<int>&	v = table[k];
	vector<vector<double>>	new_X;
	for(size_t i = 0; i < X.size(); ++i) {
		vector<double>	row;
		for(size_t j = 0; j < X[i].size(); ++j) {
			if(v[j] == g)
				row.push_back(X[i][j]);
		}
		new_X.push_back(row);
	}
	return new_X;
}

void CrossValidation::set_values(vector<double>& pred,
								const vector<double>& p_sub, size_t k, int g) {
	const vector<int>&	v = table[k];
	size_t	j = 0;
	for(size_t i = 0; i < v.size(); ++i) {
		if(v[i] == g) {
			pred[i] = p_sub[j];
			++j;
		}
	}
}

void CrossValidation::write(const string& path) const {
	ofstream	ofs(path.c_str());
	if(!ofs) {
		cerr << path << '.' << endl;
		cerr << "error : can't open " << path << '.' << endl;
		exit(1);
	}
	
	const size_t	nrep = this->num_repeats();
	for(size_t j = 1; j <= nrep; ++j)
		ofs << ",CV" << j;
	ofs << endl;
	for(size_t i = 0; i < samples.size(); ++i) {
		ofs << samples[i];
		for(size_t j = 1; j <= nrep; ++j)
			ofs << ',' << table[i][j];
		ofs << endl;
	}
}

CrossValidation CrossValidation::read(const string& path) {
	const auto		table = Common::read_csv(path);
	const size_t	H = table.size();
	const size_t	W = table.front().size();
	vector<string>	samples(H-1);
	vector<vector<int>>	matrix(W-1, vector<int>(H-1));
	for(size_t j = 0; j < H-1; ++j) {
		samples[j] = table[j+1][0];
		for(size_t i = 0; i < W-1; ++i)
			matrix[i][j] = stoi(table[j+1][i+1]);
	}
	return CrossValidation(samples, matrix);
}
